Add new class `NamedComponentParticleContainer`

RemiLehe opened the pull request:
This PR adds a new class `NamedComponentParticleContainer` from which both `PinnedMemoryParticleContainer` and `WarpXParticleContainer` now derive.

The new class has some members and member functions for accessing particle components with names (strings) instead of integers. These members and member functions used to be in `WarpXParticleContainer` ; the aim of this PR is to make them accessible to `PinnedMemoryParticleContainer` - for diagnostic purposes.

RemiLehe commented on Source/Particles/ParticleBoundaryBuffer.cpp:
```diff
@@ -172,7 +172,7 @@ void ParticleBoundaryBuffer::gatherParticles (MultiParticleContainer& mypc,
                 {
                     buffer[i] = pc.make_alike<amrex::PinnedArenaAllocator>();
-                    buffer[i].AddIntComp(false);  // for timestamp
+                    buffer[i].AddIntComp("timestamp", false);
```
We can now use the function that gives a name to this quantity.

RemiLehe commented on Source/Particles/ParticleBoundaryBuffer.cpp:
```diff
@@ -215,7 +215,7 @@ void ParticleBoundaryBuffer::gatherParticles (MultiParticleContainer& mypc,
         {
             buffer[i] = pc.make_alike<amrex::PinnedArenaAllocator>();
-            buffer[i].AddIntComp(false);  // for timestamp
+            buffer[i].AddIntComp("timestamp", false);
```
We can now use the function that gives a name to this quantity.

RemiLehe commented on Source/Particles/ParticleBoundaryBuffer.cpp:
```diff
@@ -273,7 +273,7 @@ int ParticleBoundaryBuffer::getNumParticlesInContainer(
 }
 
-WarpXParticleContainer::ContainerLike<amrex::PinnedArenaAllocator> &
+PinnedMemoryParticleContainer &
```
The corresponding `.H` file indicates that this function should return a `PinnedMemoryParticleContainer`.
Now that `PinnedMemoryParticleContainer` is a type of `NamedComponentParticleContainer`, it is not equivalent anymore to use `WarpXParticleContainer::ContainerLike<amrex::PinnedArenaAllocator>` instead of `PinnedMemoryParticleContainer`.

ax3l replied:
Should we move `ContainerLike` from `WarpXParticleContainer` up to `NamedComponentParticleContainer` now?

RemiLehe replied:
`ContainerLike` is actually a type/function that is defined in `amrex::ParticleContainer`, I think, and then `WarpXParticleContainer` inherits it.

ax3l replied:
Oh, I added it there, hah. Overlooked, sorry :)

RemiLehe commented on Source/Particles/PinnedMemoryParticleContainer.H:
```diff
@@ -1,14 +1,8 @@
-#include <AMReX_ParIter.H>
-
-using PinnedMemoryParticleContainer = WarpXParticleContainer::ContainerLike<amrex::PinnedArenaAllocator>;
+using PinnedMemoryParticleContainer = NamedComponentParticleContainer<amrex::PinnedArenaAllocator>;
```
`PinnedMemoryParticleContainer` is now a kind of `NamedComponentParticleContainer`.

RemiLehe commented on Source/Particles/WarpXParticleContainer.H:
```diff
@@ -111,7 +111,7 @@ public:
  */
 class WarpXParticleContainer
-    : public amrex::ParticleContainer<0,0,PIdx::nattribs>
+    : public NamedComponentParticleContainer<amrex::DefaultAllocator>
```
`WarpXParticleContainer` now derives from `NamedComponentParticleContainer`.

RemiLehe commented on Source/Particles/WarpXParticleContainer.H:
```diff
@@ -289,35 +289,6 @@ public:
     // split along diagonals (0) or axes (1)
     int split_type = 0;
 
```
The functions below were moved to `NamedParticleContainer`.

RemiLehe commented on Source/Particles/WarpXParticleContainer.H:
```diff
@@ -379,11 +345,6 @@ public:
      void defineAllParticleTiles () noexcept;
 
 protected:
```
The member variables below were moved to `NamedComponentParticleContainer`.

RemiLehe commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,120 @@
+#include <AMReX_ParIter.H>
+#include "Utils/TextMsg.H"
+
+struct PIdx
```
This `struct` used to be in `WarpXParticleContainer_fwd.H`.

ax3l replied:
Generally speaking, I would like to move this into its own file in WarpX.

That way, we could move `NamedComponentParticleContainer` into ABLASTR.

RemiLehe commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,127 @@
+#ifndef NamedComponentParticleContainer_H_
```
@ax3l 
Here, should we split this file into a `_fwd.H` file and a regular `.H` file?

ax3l replied:
yes, if you have a use case where you can already use `NamedComponentParticleContainer<T>` without accessing its members?

Being able to use `PIdx` on its own is already worth it, I think.

RemiLehe commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+class NamedComponentParticleContainer :
+public amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>
+{
+public:
```
Most of the functions below where moved over from `WarpXParticleContainer`

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+#ifndef NamedComponentParticleContainer_H_
```
Please add copyright header :)

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+ *  with a name (string) instead of doing so with an integer index.
+ *
+ *  This is done by storing maps that give the index of the component
+ *  that corresponds to a given string.
```
please add `@tparam T_Allocator ...`

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+ *  This is done by storing maps that give the index of the component
+ *  that corresponds to a given string.
+ */
+template <template<class> class Allocator>
```
maybe rename to `T_Allocator` to avoid potential naming collisions.

ax3l replied:
maybe add `= amrex::DefaultAllocator` default?

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+        particle_comps["theta"] = PIdx::theta;
+    #endif
+    }
+    virtual ~NamedComponentParticleContainer() {}
```
```suggestion

    virtual ~NamedComponentParticleContainer() = default;
```

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    std::map<std::string, int> getParticleComps () const noexcept { return particle_comps;}
```
Do you need sorted keys? Otherwise please use `std::unordered_map`

RemiLehe replied:
OK, I tried that, but this breaks the code. This is because these functions are **already** used in the rest of the code - since this PR is just moving them from `WarpXParticleContainer` to `NamedParticleContainer` - and the part of the code that calls the function expects a `map`.

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::AddRealComp;
+    using amrex::ParticleContainer<0,0,PIdx::nattribs,0,Allocator>::AddIntComp;
+
+    void AddRealComp (const std::string& name, bool comm=true)
```
Please add doc strings to all methods, including getters :)
Doc strings should explain the abbreviations we use in the names.

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    std::map<std::string, int> getParticleComps () const noexcept { return particle_comps;}
```
Consider to clarify in the  doc string:
- components means SoA or AoS or both?
- is `Real` implicit?
- can we use `Int` instead if `i`?
- What is different for "Runtime"?

RemiLehe replied:
OK, I will add the docstrings. 
Regarding the naming of the functions: note that this code was moved from `WarpXParticleContainer`, hence the names are the same as the previous version of these functions. We could potentially change the names, but that would imply changing many places in the code where we are **already** using these functions. This could be done in a second PR.

ax3l replied:
Sounds good, yes that explains it.
Let's refactor in a follow-up.

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+    NamedComponentParticleContainer ( NamedComponentParticleContainer && ) = default;
+    NamedComponentParticleContainer& operator= ( NamedComponentParticleContainer && ) = default;
+
+    /** Create an empty particle container
```
Consider adding the `ContainerLike` trait here as well? (Move up from `WarpXParticleContainer`?)

RemiLehe replied:
Hm, actually, I think that `ContainerLike` is a trait of the parent class, i.e. `amrex::ParticleContainer`

ax3l replied:
I forgot I added it upstream, thx!

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,128 @@
+#include <AMReX.H>
+#include <AMReX_AmrParGDB.H>
+#include <AMReX_Particles.H>
+#include "Utils/TextMsg.H"
```
- include WarpX includes before AMReX
- Includes missing for classes used from stdlib

RemiLehe commented:
Thanks a lot for your review @ax3l, and sorry for the late update.
I updated the code (and in particular added docstrings), and addressed some of the comments directly on Github.

ax3l commented on Source/Particles/NamedComponentParticleContainer.H:
```diff
@@ -0,0 +1,178 @@
+    std::map<std::string, int> getParticleRuntimeiComps () const noexcept { return particle_runtime_icomps;}
+
+protected:
+    std::map<std::string, int> particle_comps;
```
I think these should all be the faster `std::unordered_map`s, because we don't care about the order of the `key` (`std::string`) values in the container:
https://en.cppreference.com/w/cpp/container/unordered_map

ax3l reviewed (approved):
Good for now, leftover open comments can go into follow-up PRs.
